core/storage/suggestion/gae_models.py

```diff
@@ -0,0 +1,267 @@
+# limitations under the License.
+
+"""Models for Oppia suggestions."""
+
+from core.platform import models
+import feconf
+
+from google.appengine.ext import ndb
+
+(base_models,) = models.Registry.import_models([models.NAMES.base_model])
+
+# Constants defining types of entities to which suggestions can be created.
+ENTITY_TYPE_EXPLORATION = 'exploration'
+ENTITY_TYPE_QUESTION = 'question'
+ENTITY_TYPE_SKILL = 'skill'
+ENTITY_TYPE_TOPIC = 'topic'
+
+ENTITY_TYPE_CHOICES = [
+    ENTITY_TYPE_EXPLORATION,
+    ENTITY_TYPE_QUESTION,
+    ENTITY_TYPE_SKILL,
+    ENTITY_TYPE_TOPIC
+]
+
+# Constants defining the different possible statuses of a suggestion.
+STATUS_ACCEPTED = 'accepted'
+STATUS_IN_REVIEW = 'review'
+STATUS_INVALID = 'invalid'
+STATUS_REJECTED = 'rejected'
+
+STATUS_CHOICES = [
+    STATUS_ACCEPTED,
+    STATUS_IN_REVIEW,
+    STATUS_INVALID,
+    STATUS_REJECTED
+]
+
+# Constants defining various suggestion types.
+SUGGESTION_EDIT_STATE_CONTENT = 'edit_exploration_state_content'
+
+SUGGESTION_TYPE_CHOICES = [
+    SUGGESTION_EDIT_STATE_CONTENT
+]
+
+# Defines what is the minimum role required to review suggestions
+# of a particular type.
+SUGGESTION_MINIMUM_ROLE_FOR_REVIEW = {
+    SUGGESTION_EDIT_STATE_CONTENT: feconf.ROLE_ID_EXPLORATION_EDITOR
+}
+
+# Constants defining various contribution types.
+CONTRIBUTION_TYPE_CONTENT = 'content'
+CONTRIBUTION_TYPE_TRANSLATION = 'translation'
+
+CONTRIBUTION_TYPE_CHOICES = [
+    CONTRIBUTION_TYPE_CONTENT,
+    CONTRIBUTION_TYPE_TRANSLATION
+]
+
+#The delimiter to be used in the suggestion ID.
+SUGGESTION_ID_DELIMITER = '.'
+
+
+class SuggestionParams(ndb.Model):
+    """Model to store all the suggestion params for a suggestion."""
+
+    # The contribution type.
+    contribution_type = ndb.StringProperty(
+        required=True, choices=CONTRIBUTION_TYPE_CHOICES)
+    # The contribution domain. This parameter along with the above contribution
+    # type will be used to determine the category to 'score' the suggester.
+    contribution_domain = ndb.StringProperty(required=True)
+    # The ID of the entity being suggested to.
+    entity_id = ndb.StringProperty(required=True)
+    # The version number of the entity being suggested to.
+    entity_version = ndb.IntegerProperty(required=True)
+
+
+class SuggestionModel(base_models.BaseModel):
+    """Model to store suggestions made by Oppia users."""
+
+    # The type of suggestion.
+    suggestion_type = ndb.StringProperty(
+        required=True, indexed=True, choices=SUGGESTION_TYPE_CHOICES)
+    # The type of entity which the suggestion is linked to.
+    entity_type = ndb.StringProperty(
+        required=True, indexed=True, choices=ENTITY_TYPE_CHOICES)
+    # Status of the suggestion.
+    status = ndb.StringProperty(
+        required=True, indexed=True, choices=STATUS_CHOICES)
+    # Additional parameters to be stored.
+    suggestion_params = ndb.StructuredProperty(SuggestionParams)
+    # The author of the suggestion.
```

Comment:
The user ID of the author of ...

Ditto below, for reviewer_id, thread_id, etc. In general, be careful to distinguish between IDs and the objects they point to, otherwise you set yourself up for a load of confusion later.

Reply:
Done.